[autosubmit] Authenticate user

KristinBi opened the pull request:
This PR is to do the authentication on check_pull_request API.
### Future
Add authentication for webhook API.

CaseyHillers commented on auto_submit/bin/server.dart:
```diff
@@ -25,14 +26,15 @@ Future<void> main() async {
             config: config,
           ).post)
-      ..get('/check-pull-request', CheckPullRequest(config: config).get);
+      ..get('/check-pull-request', CheckPullRequest(config: config, authenticationProvider: authProvider).get);
```
nit: add a trailing comma

KristinBi replied:
Done.

CaseyHillers commented on auto_submit/lib/model/appengine/commit.dart:
```diff
@@ -0,0 +1,120 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
```
rm? Seems like you pulled all the models in which is what we want to avoid. Please only copy the minimum necessary

KristinBi replied:
rm

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,177 @@
+      throw const Unauthenticated('Invalid ID token');
+    }
+
+    // if (token.hostedDomain != 'google.com') {
```
Can you rm the dead code from this file?

KristinBi replied:
I'm so sorry for this. I added too many files and codes which made me felt overwhelmed, thus forgot this.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,177 @@
+///     runtime if the request originated from anything other than a cron job.
+///     Thus, the header is safe to trust as an authentication indicator.
+///
+///  2. If the request has the `'X-Flutter-IdToken'` HTTP header
```
This won't be supported for auto submit

KristinBi replied:
Hi Casey, the previous logic check two headers: 'X-Appengine-Cron' and 'X-Flutter-IdToken'.
Do you mean that here for autosubmit, we only need to check 'X-Appengine-Cron' and we don't need to take care about the 'X-Flutter-IdToken'?

CaseyHillers replied:
Correct. X-Flutter-IdToken is for Google accounts

KristinBi replied:
Gotcha Casey. Since we don't need to check for the Google accounts, I also don't need to get the OAuth client Id. I will also delete all logics for the OAuth client Id in Config.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,177 @@
+    if (isCron) {
+      // Authenticate cron requests
+      return AuthenticatedContext(clientContext: clientContext);
+    } else if (idTokenFromHeader != null) {
```
You won't need to support id token from header

CaseyHillers commented on auto_submit/lib/requests/check_pull_request.dart:
```diff
@@ -27,16 +28,18 @@ const int _kMergeCountPerRepo = 1;
 /// For pull requests where an 'autosubmit' label was added in pubsub,
 /// check if the pull request is mergable.
-class CheckPullRequest extends RequestHandler {
+class CheckPullRequest extends ApiRequestHandler {
```
Instead of `ApiRequestHandler` make this `AuthenticatedRequestHandler`

KristinBi replied:
Changed the file and class name.

CaseyHillers commented on auto_submit/lib/server/api_request_handler.dart:
```diff
@@ -0,0 +1,106 @@
+  /// Throws a [BadRequestException] if any of [requiredParameters] is missing
+  /// from [requestData].
+  @protected
+  void checkRequiredParameters(List<String> requiredParameters) {
```
nit: rm what isn't used

KristinBi replied:
Deleted unused logic in api_request_handler.

CaseyHillers commented on auto_submit/lib/server/request_handler.dart:
```diff
@@ -25,4 +27,79 @@ abstract class RequestHandler {
+  /// If this is called outside the context of an HTTP request, this will
+  /// throw a [StateError].
+  @protected
+  U? getValue<U>(RequestKey<U> key, {bool allowNull = false}) {
```
We should avoid pulling this code in as it adds a lot of complexity

KristinBi replied:
Deleted unused logic in request_handler.

CaseyHillers reviewed (commented):
Can you go through and remove all the code that isn't used? In it's current state, this PR adds much more complexity to the app than is necessary

KristinBi commented:
> Can you go through and remove all the code that isn't used? In it's current state, this PR adds much more complexity to the app than is necessary

Sure Casey, I went through all the codes and removed the unnecessary codes to clear the logic. I also removed codes related to X-Flutter-IdToken.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,97 @@
+///
+///  * [AuthenticationProvider]
+@immutable
+class AuthenticatedContext {
```
We don't need AuthenticatedContext or ClientContext since there's no human actions

KristinBi replied:
Gotcha, I removed both.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,97 @@
+///
+///  * <https://cloud.google.com/appengine/docs/standard/python/reference/request-response-headers>
+@immutable
+class AuthenticationProvider {
```
Since this isn't generic authentication, you should call it something like `CronAuthProvider`

KristinBi replied:
Sure, changed the class name to CronAuthProvider.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,57 @@
+///
+///
+/// If none of the above authentication methods yield an authenticated
+/// request, then the request is unauthenticated, and any call to
```
This paragraph should be reworded

KristinBi replied:
Deleted it.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,57 @@
+
+  /// Authenticates the specified [request].
+  ///
+  /// See the class documentation on [CronAuthProviderer] for a discussion
```
This paragraph isn't relevant anymore

KristinBi replied:
Dleted.

KristinBi commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -33,65 +30,28 @@ import '../service/config.dart';
-    this.httpClientProvider = Providers.freshHttpClient,
-  });
+class CronAuthProvider {
+  const CronAuthProvider(this.config);
```
Removed donfig.

CaseyHillers commented on auto_submit/lib/request_handling/authentication.dart:
```diff
@@ -0,0 +1,43 @@
+///
+///  * <https://cloud.google.com/appengine/docs/standard/python/reference/request-response-headers>
+@immutable
+class CronAuthProvider {
```
nit: You should leave a TODO to generalize this to implement from a `AuthProvider` that defines it implements `authenticate()`

KristinBi replied:
Added a TODO. Thank you very much Casey for helping review my PR.
